### DISCUSSION TITLE
addSelect / selectRaw override 'select *'

### DISCUSSION THREAD

**Discussion by scibuff** [laravel/framework#36113] [Q&A]
- Laravel Version: 8.19.0
- PHP Version: 7.4.11
- Database Driver & Version: mysql 8.0.21-0ubuntu0.20.04.4 for Linux on x86_64 ((Ubuntu))

### Description:

When I try to add a custom column to my queries with `addSelect` or `selectRaw` only the specified column is returned if 
`$builder->columns` is `null`

### Steps To Reproduce:

```php

class Product extends Model 
{
    protected $fillable = ['name', 'cost', 'shipping'];

    protected static function boot()
    {
        parent::boot();

        static::addGlobalScope('total', function(Builder $builder)
        {            
            $builder->addSelect( DB::Raw('cost + shipping AS total') );
//            $builder->selectRaw( 'cost + shipping AS total' ); // same result as the above
        }
    }
}

...
    dump( Product::all()->toArray() );

>>>
array:1 [
  0 => array:1 [
    "total" => "10"
  ]
]
```

### Exptected:

```php
...
    dump( Product::all()->toArray() );

>>>
array:1 [
  0 => array:1 [
    "id" => "1" 
    "cost" => "9.5"
    "shipping" => "0.5"
    "total" => "10"
  ]
]
```

Looking at the `Builder->addSelect` I see I need to do `$builder->select( $builder->from.'.*' );` in my `addGlobalScope` closure, but this should be either specified in the docs (currently it says to use `addSelect`)

> If your global scope is adding columns to the select clause of the query, you should use the `addSelect` method instead of `select`. This will prevent the unintentional replacement of the query's existing select clause.

or the `addSelect` should do it for non 'queryable` columns as well

**Comment by veelasky** [comment]
why not using `get` attribute [accessor](https://laravel.com/docs/8.x/eloquent-mutators#defining-an-accessor) ?!


EDIT: let me elaborate

```php
class Product 
{
    public function getTotalAttribute()
    {
         return $this->cost + $this->total;
    }
}
```


if you want to add to every `array` representation:
```php
class Product 
{
    protected $appends = [ 'total' ];

    // .... the rest
}
```

**Comment by scibuff** [comment]
Thank you, I am aware of the `accessors` and `append` option in the `Model`. However, the point of adding the select statement is the ability to then use the calculated column in a query (e.g. get the cheapet/most expensive products), which is not possible with appended model attributes.

The example I posted is a simple one but there certainly are more complicated calculations one may need to use for  `orderBy`

**Comment by BrandonSurowiec** [comment]
@scibuff In a scope you can do this:

```php
function scopeWithTotal($query) {

    if (is_null($query->getQuery()->columns)) {
        $query->select($this->qualifyColumn('*'));
    }

    $query->addSelect(DB::raw('cost + shipping AS total'));
}
```

Then:

```php
        static::addGlobalScope('total', function(Builder $builder) {            
            $builder->withTotal();
        }
```

Or you can just be explicit about calling `withTotal()` in whatever queries need it. I do find when combining a bunch of dynamic select queries like this, sometimes the order matters, but I don't think you'll run into that here.

**Reply by scibuff** [reply]
Thank you! That is exactly what I am doing, but my point is that if I use `query->addSelect` I shouldnt need to add `$query->select('*')` as the `addSelect` is not supposed to overwrite the current/default select (which is to select *). The documentation specifically states this, but it is not the implemented behavior.

**Reply by BrandonSurowiec** [reply]
Agree, but it seems it is here to stay at the moment:
https://github.com/laravel/framework/pull/29588
https://github.com/laravel/framework/pull/24436

Maybe it is something that could get merged into 9.x as it would be a breaking change, but I don't know if Taylor has changed his stance on this.

[Edit: Especially considering this was originally an "issue", which Taylor locked and moved to a discussion.]

**Comment by vHeemstra** [comment]
I agree this is (years later) still weird behaviour. Even if it is a breaking change, it should be fixed.

**The `addSelect` method now has very inconsistent handling:**

- If you `addSelect()` using an associative array (so selecting using aliases), the `table.*` **IS** added before added your new select.

https://github.com/laravel/framework/blob/5a7f2b4742b3dc7ce43acc698f400a9395801c7b/src/Illuminate/Database/Query/Builder.php#L417-L421

- If you `addSelect()` using string or numerical array, the `table.*` **IS NOT** added (thus losing the all columns selector).

https://github.com/laravel/framework/blob/5a7f2b4742b3dc7ce43acc698f400a9395801c7b/src/Illuminate/Database/Query/Builder.php#L427

**Comment by RichardStyles** [comment]
Some further notes, as I was looking to add a scope which adds a column ad-hoc, that relies on a sub query.

It seems the issue is arisen from the `$this->isQueryable($column)` check.
https://github.com/laravel/framework/blob/5050195f1841f1f07be4f1a61a2906c32afb67ad/src/Illuminate/Database/Query/Builder.php#L416
Added that `DB::raw(...)` is not regarded as querable.
https://github.com/laravel/framework/blob/5050195f1841f1f07be4f1a61a2906c32afb67ad/src/Illuminate/Database/Query/Builder.php#L3995-L4001

**Comment by RomainMazB** [comment]
Still a thing in 2025, here is how I personally tackled this:
```php
public function scopeWithIsLiked(Builder $query): Builder
{
    $subQuery = /* Useless for the example */;

    // Avoid to override current query columns select
    $selects = $query->getSelect() ?? ['*'];
    array_push($selects, DB::raw("exists($subQuery) as is_liked"));

    return $query->select($selects);
}
```

**Comment by f-liva** [comment]
For those interested, I created a small Laravel package that introduces the a `mergeSelect` method on Eloquent and Query builders and solves the problem:

https://github.com/f-liva/eloquent-merge-select
